### ISSUE TITLE
Bug:  `partitionByComment` Doesn't Respect Comment Boundaries in `sort-union-types`

### ISSUE THREAD

**Issue by akibarika** [azat-io/eslint-plugin-perfectionist#306]
### Describe the bug

When using the `partitionByComment: true` option in `sort-union-types`, the union type members are expected to be sorted within groups that are separated by comments. However, the current behavior ignores comment boundaries and sorts the entire union type globally, even across different comment sections.

### Code example

Given this input:

```tsx
type T =
  // Group 1
  | Z
  | B
  // Group 2
  | C
  | A;
```

Expected output (sorting within comment partitions):

```
type T =
  // Group 1
  | B
  | Z
  // Group 2
  | A
  | C;
```

Actual output (comments ignored, globally sorted):

```
type T =
  // Group 1
  | A
  | B
  // Group 2
  | C
  | Z;
```

### ESLint version

8.57.0

### ESLint Plugin Perfectionist version

3.8.0

### Additional comments

`partitionByComment` works in others, for example `sort-exports`

### Validations

- [X] Read the [docs](https://perfectionist.dev).
- [X] Check that there isn't [already an issue](https://github.com/azat-io/eslint-plugin-perfectionist/issues) that reports the same bug to avoid creating a duplicate.

**Comment by azat-io** [comment]
Thank you for your issue!
I was able to reproduce the problem. The problem seems to be in this file: https://github.com/azat-io/eslint-plugin-perfectionist/blob/main/utils/get-node-range.ts

**Comment by hugop95** [comment]
Hello @akibarika cc @azat-io ,

This problem appears on `sort-intersection-types` as well and is mentioned in the PRs for both:
https://github.com/azat-io/eslint-plugin-perfectionist/pull/286
https://github.com/azat-io/eslint-plugin-perfectionist/pull/287

The main reason comes from how `getCommentsBefore` work:
it uses the `getCommentsBefore(nodeOrToken: TSESTree.Node | TSESTree.Token): TSESTree.Comment[];` function from `eslint`.

https://github.com/azat-io/eslint-plugin-perfectionist/blob/4e7e5ad1c5635f677288fc192f1168e611da584a/utils/get-node-range.ts#L47

When you write

```
  // Group 1
  | Z
```

`|` is considered a token, so `// Group 1` is not considered to be a comment associated to the token `Z`.

```
  // Group 1
  Z |
```
on the other hand, will work.

We will have to wrap something around this function to bypass this case (and `&` as well for intersection types) if we want to solve this.

**Comment by azat-io** [comment]
Fixed in e5279ac6a04e9b0463ccb606e440f50e48447933.
Released in [v4.0.0](https://github.com/azat-io/eslint-plugin-perfectionist/releases/tag/v4.0.0).

If you'd like, you can support the release with a retweet:
https://x.com/azat_io_en/status/1858853714238271530
